### PR TITLE
 Convert MockResponse to Kotlin

### PR DESCRIPTION
**Pull request by TangHuaiZhe**

### CODE REVIEW

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+ */
+package okhttp3.mockwebserver
+
+import okhttp3.Headers
+import okhttp3.WebSocketListener
+import okhttp3.internal.addHeaderLenient
+import okhttp3.internal.http2.Settings
+import okhttp3.mockwebserver.internal.duplex.DuplexResponseBody
+import okio.Buffer
+import java.util.ArrayList
```
Comment (JakeWharton) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Delete import

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+  /**
+   * Sets the response body to [body], chunked every [maxChunkSize] bytes.
+   */
+  fun setChunkedBody(body: Buffer, maxChunkSize: Int): MockResponse {
+    removeHeader("Content-Length")
+    headers.add(CHUNKED_BODY_HEADER)
+
+    val bytesOut = Buffer()
+    while (!body.exhausted()) {
+      val chunkSize = Math.min(body.size, maxChunkSize.toLong())
```
Comment (JakeWharton) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Kotlin's `minOf`

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java [hunk 3]
```diff
@@ -106,7 +106,7 @@
     MwsDuplexAccess.instance = new MwsDuplexAccess() {
       @Override public void setBody(
           MockResponse mockResponse, DuplexResponseBody duplexResponseBody) {
-        mockResponse.setBody(duplexResponseBody);
+        mockResponse.setBody$mockwebserver(duplexResponseBody);
```
Comment (JakeWharton) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java]: I'm not sure how safe this is to rely on...

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    }
+  }
+
+  /** Returns the HTTP response line, such as "HTTP/1.1 200 OK".  */
+  fun getStatus(): String {
+    return status
+  }
+
+  fun setResponseCode(code: Int): MockResponse {
+    var reason = "Mock Response"
```
[the remark this answers is not in the document]
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Suddenly see the light,Fixed

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+
+/** A scripted response to be replayed by the mock web server.  */
+class MockResponse : Cloneable {
+  private var status: String = ""
+  private var headers = Headers.Builder()
+  private var trailers = Headers.Builder()
+
+  private var body: Buffer = Buffer()
+
+  var throttleBytesPerPeriod = java.lang.Long.MAX_VALUE
```
[the remark this answers is not in the document]
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 1 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 2 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java (hunk 3 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java]: It does seem to be unstable, `./gradlew clean build `can always pass, but IDEA occasionally can't find references. I removed all the `internal` in this class and fixed the reference.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    setHeader("Content-Length", 0)
+  }
+
+  public override fun clone(): MockResponse {
+    try {
+      val result = super.clone() as MockResponse
+      result.headers = headers.build().newBuilder()
+      result.promises = ArrayList(promises)
+      return result
+    } catch (e: CloneNotSupportedException) {
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: you can omit this try/catch clause

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 7]
```diff
@@ -13,330 +13,307 @@
...
+class MockResponse : Cloneable {
+  private var status: String = ""
+  private var headers = Headers.Builder()
+  private var trailers = Headers.Builder()
+
+  private var body: Buffer? = null
+
+  var throttleBytesPerPeriod = Long.MAX_VALUE
+    private set
+  private var throttlePeriodAmount: Long = 1
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: `private var throttlePeriodAmount = 1L`

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 8]
```diff
@@ -13,330 +13,307 @@
...
+  private var throttlePeriodUnit = TimeUnit.SECONDS
+
+  private var socketPolicy = SocketPolicy.KEEP_OPEN
+  private var http2ErrorCode = -1
+
+  private var bodyDelayAmount: Long = 0
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: `private var bodyDelayAmount = 0L`

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 9]
```diff
@@ -13,330 +13,307 @@
...
+  private var bodyDelayUnit = TimeUnit.MILLISECONDS
+
+  private var headersDelayAmount: Long = 0
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: `private var headersDelayAmount = 0L`

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 10]
```diff
@@ -13,330 +13,307 @@
...
+  private var headersDelayUnit = TimeUnit.MILLISECONDS
+
+  private var promises: MutableList<PushPromise> = ArrayList()
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: `private var promises = mutableListOf<PushPromise>()`

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 11]
```diff
@@ -13,330 +13,307 @@
...
+    get() = duplexResponseBody != null
+
+  /** Returns the streams the server will push with this response.  */
+  val pushPromises: List<PushPromise>
+    get() = promises
+
+  /** Creates a new mock response with an empty body.  */
+  init {
+    setResponseCode(200)
+    setHeader("Content-Length", 0)
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: `setHeader("Content-Length", 0L)`

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 12]
```diff
@@ -13,330 +13,307 @@
...
     try {
-      MockResponse result = (MockResponse) super.clone();
-      result.headers = headers.build().newBuilder();
-      result.promises = new ArrayList<>(promises);
-      return result;
-    } catch (CloneNotSupportedException e) {
-      throw new AssertionError();
+      val result = super.clone() as MockResponse
+      result.headers = headers.build().newBuilder()
+      result.promises = ArrayList(promises)
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: `result.promises = promises.toMutableList()`

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 13]
```diff
@@ -13,330 +13,307 @@
...
+      throw AssertionError()
     }
   }
 
-  /** Returns the HTTP response line, such as "HTTP/1.1 200 OK". */
-  public String getStatus() {
-    return status;
+  /** Returns the HTTP response line, such as "HTTP/1.1 200 OK".  */
+  fun getStatus(): String {
+    return status
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: expression function

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 14]
```diff
@@ -13,330 +13,307 @@
...
-    return headers.build();
+  /** Returns the HTTP headers, such as "Content-Length: 0".  */
+  fun getHeaders(): Headers {
+    return headers.build()
   }
 
-  public Headers getTrailers() {
-    return trailers.build();
+  fun getTrailers(): Headers {
+    return trailers.build()
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: expression function

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 15]
```diff
@@ -13,330 +13,307 @@
...
+  fun removeHeader(name: String): MockResponse {
+    headers.removeAll(name)
+    return this
   }
 
-  DuplexResponseBody getDuplexResponseBody() {
-    return duplexResponseBody;
+  /** Returns a copy of the raw HTTP payload.  */
+  fun getBody(): Buffer? {
+    return body?.clone()
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: nice!

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 16]
```diff
@@ -13,330 +13,307 @@
...
-    this.socketPolicy = socketPolicy;
-    return this;
+  fun setSocketPolicy(socketPolicy: SocketPolicy): MockResponse {
+    this.socketPolicy = socketPolicy
+    return this
   }
 
-  public int getHttp2ErrorCode() {
-    return http2ErrorCode;
+  fun getHttp2ErrorCode(): Int {
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: expression function

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 17]
```diff
@@ -13,330 +13,307 @@
...
-  public long getHeadersDelay(TimeUnit unit) {
-    return unit.convert(headersDelayAmount, headersDelayUnit);
+  fun getHeadersDelay(unit: TimeUnit): Long {
+    return unit.convert(headersDelayAmount, headersDelayUnit)
   }
 
   /**
-   * When {@link MockWebServer#setProtocols(java.util.List) protocols} include {@linkplain
-   * okhttp3.Protocol#HTTP_2}, this attaches a pushed stream to this response.
+   * When [protocols][MockWebServer.setProtocols] include [HTTP_2][okhttp3.Protocol], this attaches a pushed stream to this response.
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: nit: wrap at 100 columns

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt [hunk 18]
```diff
@@ -13,330 +13,307 @@
...
-    return promises;
+  fun withPush(promise: PushPromise): MockResponse {
+    this.promises.add(promise)
+    return this
   }
 
   /**
-   * When {@linkplain MockWebServer#setProtocols(java.util.List) protocols} include {@linkplain
-   * okhttp3.Protocol#HTTP_2 HTTP/2}, this pushes {@code settings} before writing the response.
+   * When [protocols][MockWebServer.setProtocols] include [HTTP_2][okhttp3.Protocol], this pushes [settings] before writing the response.
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: wrap at 100 columns

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 6 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 7 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 8 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 9 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 10 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 11 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 12 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 13 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 14 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 16 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 17 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt (hunk 18 above)
Comment (TangHuaiZhe) [reply mockwebserver/src/main/java/okhttp3/mockwebserver/MockResponse.kt]: Fixed.

**Pull request merged**